react/features/base/avatar/components/AbstractAvatar.js

```diff
@@ -0,0 +1,215 @@
+     * the background color match the string that the initials are generated from.
+     */
+    colorBase?: string,
+
+    /**
+     * Display name of the entity to render an avatar for (if any). This is handy when we need
+     * an avatar for a non-participasnt entity (e.g. a recent list item).
+     */
+    displayName?: string,
+
+    /**
+     * The ID of the participant to render an avatar for (if it's a participant avatar).
+     */
+    participantId?: string,
+
+    /**
+     * The size of the avatar.
+     */
+    size: number,
+
+    /**
+     * URI of the avatar, if any.
+     */
+    uri: ?string,
+}
+
+type State = {
+    avatarFailed: boolean,
+    gravatarFailed: boolean
+}
+
+export const DEFAULT_SIZE = 65;
+
+/**
+ * Implements an abstract class to render avatars in the app.
+ */
+export default class AbstractAvatar<P: Props> extends PureComponent<P, State> {
+    /**
+     * Instantiates a new {@code Component}.
+     *
+     * @inheritdoc
+     */
+    constructor(props: P) {
+        super(props);
+
+        this.state = {
+            avatarFailed: false,
+            gravatarFailed: false
+        };
+
+        this._onAvatarLoadError = this._onAvatarLoadError.bind(this);
+        this._onGravatarLoadError = this._onGravatarLoadError.bind(this);
+    }
+
+    /**
+     * Implements {@code Component#componentDidUpdate}.
+     *
+     * @inheritdoc
+     */
+    componentDidUpdate(prevProps: P) {
+        if (prevProps._avatarURL !== this.props._avatarURL
+            || prevProps._gravatarURL !== this.props._gravatarURL) {
+
+            // URLS changed, so we need to try to fetch them again.
+            // Eslint doesn't like this statement, but based on the React doc, it's safe if it's
+            // wrapped in a condition: https://reactjs.org/docs/react-component.html#componentdidupdate
+
+            // eslint-disable-next-line react/no-did-update-set-state
+            this.setState({
+                avatarFailed: false,
+                gravatarFailed: false
+            });
+        }
+    }
+
+    /**
+     * Implements {@code Componenr#render}.
+     *
+     * @inheritdoc
+     */
+    render() {
+        const {
+            _avatarURL,
+            _gravatarURL,
+            _initialsBase,
+            colorBase
+        } = this.props;
+        const {
+            avatarFailed,
+            gravatarFailed
+        } = this.state;
+
+        const _initials = getInitials(_initialsBase);
+
+        // Precendence logic of avatars:
+
+        if (_avatarURL && _avatarURL.match(/^http[s]*:\/\//i) && !avatarFailed) {
+            // URL-based avatar first, if set and succeeds to load
+            return this._renderURLAvatar(_avatarURL);
+        } else if (_gravatarURL && !gravatarFailed) {
+            // Then gravatar if there is no URL based avatar, and there is email address
+            // and gravatar succeeds to load
+            return this._renderGravatar(_gravatarURL);
```

Comment:
why are we passing `_gravatarURL` to the function? it could read it from `props` like other parameters such as the size.

Reply:
That code is now totally different.

There is still a similar call in there though. I did that because I felt it more consistent. We want to render an avatar with some details provided that may or may not be available at the place of render. In this case it is, true, but in case of the other call, it's not. So I went for function footprint consistency instead of minimalistic. I guess it's a matter of taste